### PR TITLE
Covariance Module Improvements

### PR DESCRIPTION
**Pull request by austinTalbot7241993**
We are now realizing that fitting covariance matrices is a critical task and while there is extensive work in the statistics community on improving finite-sample properties of estimation, very little has migrated to other fields. Part of this is likely due to scattered implementations across a variety of github repositories. We are implementing a large quantity of these estimators ourselves with a common format similar to sklearn. Here I implement two shrinkage estimators by Ledoit and Wolfe.

### CODE REVIEW

File: python/python/bystro/covariance/_covariance_np.py [hunk 1]
```diff
@@ -106,3 +95,146 @@ def _fill_prior_options(self, prior_options: dict[str, Any]) -> dict[str, Any]:
...
+
+    def fit(self, X: NDArray[np.float_]) -> "LinearShrinkageCovariance":
+        """
+        This fits a covariance matrix using the linear shrinkage approach 
+        to combine the empirical covariance matrix with a structured 
+        estimator.
+
+        Parameters
+        ----------
+        X : np.array-like, (n_samples, n_covariates)
```
Comment (akotlar) [inline python/python/bystro/covariance/_covariance_np.py:110]: Let's add a note that we do not support missing values

File: python/python/bystro/covariance/_covariance_np.py [hunk 2]
```diff
@@ -106,3 +95,146 @@ def _fill_prior_options(self, prior_options: dict[str, Any]) -> dict[str, Any]:
...
+    the ratio of variables p to n is significant.
+
+    The theoretical basis of this approach is built around the optimal 
+    estimation of p parameters, which balances between overfitting p^2
+    parameters and underfitting with only 1 parameter. The nonlinear shrinkage 
+    is calculated using an oracle estimator that depends on a sample spectral density 
+    function approximated via kernel estimation using the Epanechnikov kernel. 
+    This estimation facilitates the derivation of shrinkage intensities for each 
+    eigenvalue, tailoring the adjustment to improve estimations under various 
+    data conditions. Practical implementation details are provided, adapted from a
```
Comment (akotlar) [inline python/python/bystro/covariance/_covariance_np.py:166]: Let's remove this portion:

```
Practical implementation details are provided, adapted from a 
    Matlab script by the authors, ensuring that users can apply this method effectively 
    to empirical data sets. The docstring also underscores the utility of this estimator 
    in practical settings, providing a more parameter-efficient approach compared 
    to traditional methods, which often rely on unfeasibly large parameter sets.
    ```

File: python/python/bystro/covariance/_covariance_np.py (hunk 1 above)
Comment (austinTalbot7241993) [reply python/python/bystro/covariance/_covariance_np.py:110]: Where do you want to add it?

Comment (akotlar) [reply python/python/bystro/covariance/_covariance_np.py:109]: I was just thinking right here: "The centered data. Must not contain any missing values (np.nan)"

File: python/python/bystro/covariance/_covariance_np.py [hunk 3]
```diff
@@ -106,3 +95,146 @@ def _fill_prior_options(self, prior_options: dict[str, Any]) -> dict[str, Any]:
...
+        """
+        This fits a covariance matrix using the linear shrinkage approach 
+        to combine the empirical covariance matrix with a structured 
+        estimator.
+
+        Parameters
+        ----------
+        X : np.array-like, (n_samples, n_covariates)
+            The centered data
+
```
Comment (akotlar) [inline python/python/bystro/covariance/_covariance_np.py:111]: Let's also add a Raises section:

```py
Raises
------
ValueError:
     A value error will be raised if missing data is found in X ( np.isnan(X) evaluates to true ), or if X is not an NDArray
```

Comment (akotlar) [reply python/python/bystro/covariance/_covariance_np.py:111]: This follows numpy conventions: https://numpydoc.readthedocs.io/en/latest/format.html#raises

**Review by akotlar — APPROVED** [review]
LGTM

**Pull request merged**